### DISCUSSION TITLE
broadcast cast/observe problems

### DISCUSSION THREAD

**Discussion by michaelwareman** [pybricks/support#1087] [Q&A]
Hello:
I am working on a Johnny  5 robot with two hubs. I have been having problems with using the broadcast/observe feature in the new beta pybricks version.

1)
The program would hang on either the broadcast or observe when running. Sometimes even the hubs would not disconnect when this hangs. The only solution was to restart the laptop (Windows 10) and remove the battery from the hub.

I then thought that maybe there are too many messages being sent, both from the broadcast from each hub, and the print statement. So I put a wait statement (for now wait(100) is what I have set) after each of the broadcast statements for each hub. The code has been running for several minutes now. Two strange behaviors:
  1) the arms would pause for several seconds before continuing again
  2) after several minutes the arms are no longer synced, but move in opposite directions

2)
The other problem I have is that the code only works if I have the print statements enabled. The code for both hubs is listed below. Any help or suggestions would be greatly appreciated.

Mike,

```python
# Drive Hub1
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Motor
from pybricks.tools import wait
from pybricks.parameters import Port, Color

hub1 = InventorHub(broadcast_channel=2, observe_channels=[1])

right_arm_target_angle = 0
right_arm_angle = 0
left_arm_angle = 0
left_arm_target_angle = 0

hub1.light.on(Color.GREEN)

while True:
    data = hub1.ble.observe(1)

    if data is not None:
        left_arm_angle = data[0]
        right_arm_angle = data[1]

        #move arms to down ititialization position
        if left_arm_angle < 10:
            left_arm_target_angle = 300

        if right_arm_angle < 10:
            right_arm_target_angle = 300
        
        if left_arm_angle > 295:
            left_arm_target_angle = 0
        
        if right_arm_angle >295:
            right_arm_target_angle = 0

    print(left_arm_angle, right_arm_angle)
    hub1.ble.broadcast(left_arm_target_angle, right_arm_target_angle)
    wait(100)
```
```python
#Upper hub2 controls arms
from pybricks.hubs import InventorHub
from pybricks.pupdevices import Remote, Motor
from pybricks.tools import wait
from pybricks.parameters import Color, Button, Port, Direction

hub2 = InventorHub(broadcast_channel=1, observe_channels=[1, 2])
left_arm_motor = Motor(Port.E, Direction.COUNTERCLOCKWISE, [20, 60])
right_arm_motor = Motor(Port.F, Direction.CLOCKWISE, [20, 60])

hub2.light.on(Color.YELLOW)

while True:
    left_arm_target = left_arm_motor.angle()
    right_arm_target = right_arm_motor.angle()
    hub2.ble.broadcast(left_arm_target, right_arm_target)
    wait(100)
    print(left_arm_target, right_arm_target)

    data = hub2.ble.observe(2)
    

    if data is not None:
        left_arm_target = data[0]
        right_arm_target = data[1]
        #arms move one at a time
        #left_arm_motor.run_target(75, left_arm_target)
        #right_arm_motor.run_target(75, right_arm_target)
        #arms move together
        left_arm_motor.run_target(75, left_arm_target, wait=False)
        right_arm_motor.run_target(75, right_arm_target, wait=False)

        print(left_arm_target, right_arm_target)
```

**Comment by dlech** [comment]
> I have been having problems with using the broadcast/observe feature in the new beta

Thanks for trying it out.

> The only solution was to restart the laptop (Windows 10)

It would be helpful if you could log Bluetooth packets as describe at https://bleak.readthedocs.io/en/latest/troubleshooting.html#windows-10 so we can see what is going on.


> remove the battery from the hub.

This should never happen. It should always be possible to press and hold the main button to power off the hub even if it stops responding to everything else or if there was a really bad crash, the hub will reboot or power itself off.



> I then thought that maybe there are too many messages being sent,

Messages are sent every 100ms no matter how many times you call broadcast.


> The program would hang on either the broadcast or observe when running

I have run the program both with and without the prints enabled and haven't been able to reproduce the hang. Tried while connected to Linux, connected to Windows and not connected.

> Any help or suggestions would be greatly appreciated.

Instead of trying to implement motor control with feedback over the Bluetooth connection, I would just send a state that you want the arms to be in and do the motor control/feedback loop on the same hub that the motor are connected to. 

It would help if you could describe the big picture of how you actually want the robot to behave.

**Reply by michaelwareman** [reply]
I first started building this Johnny5 robot -- https://rebrickable.com/mocs/MOC-113842/iwogla/johny-5-mindsorms/#details (Not the best as the building instructions do not show the building process very well. What I liked was the use of the Spike Prime/Robot Inventor hubs. Which gives me more motors and sensors to work with There do appear to be a newer set of pdf and io files. I have not tried them out yet.) I also have this Johnny 5 building instructions -- https://rebrickable.com/mocs/MOC-125609/RJBrickBuilds/rjs-lego-johnny-5-atte-and-tx225-and-terminator-3s-t1-battlefield-robot/#details  I like the overall look of this robot so will be using it to finish my Johnny 5.

So, the base Hub1 has the drive motors, and the two motor that controls the raising and lowering of the lower and upper body parts. For now I have a touch senors to detect when the lower body part is fully down. That leaves me one extra port maybe for turning on/off lights for eyes? The back of the upper body part has Hub2 which has the left and right arm motors, and the 2 motors that control the turning and tilting of the head. In the first set of building instructions are motors for missile launchers.I have not decided what I am going to do that or something else. Once I have a basic Johnny 5 that can move, avoid obstacles, raise and lower arms, move head, and any other functions I want to create a finite state machine to give the robot a bit more personality. For now I just need to get the basic functions working between the two hubs.

From your last reply does this mean there is a new firmware to install onto the Hubs from the beta.pybricks.com application? Also, since broadcast/observe appears to be working better for me I was going the combine my drive and body movement code with my broadcast code for Hub1? But, now I have a new problem in that the Drivebase is not available in the beta firmware? I do not need the drivebase but it did help with allowing the robot move with more control. If I cannot have both broadcast/observe and Drivebase how should I best control the two drive motors (note the drive motors are two large PowerUp motors.)

**Reply by michaelwareman** [reply]
Sorry, it was late and I wanted to respond to your reply. I now have the DriveBase working in the beta.pybricks.com application. I will start combining my movement code and the broadcast/observe code and see what happens. I then need to add the head motors and program their movements. I will let you know how I progress.

Thank you for the assistance.

**Reply by dlech** [reply]
> From your last reply does this mean there is a new firmware to install onto the Hubs from the beta.pybricks.com application?

There is not a new firmware at beta.pybricks.com yet but if you want to try the latest build, you can always find it at https://nightly.link/pybricks/pybricks-micropython/workflows/build/master.

**Comment by dlech** [comment]
If both prints and waits are removed on the Hub1 program, I did get the hub to crash (powers off unless connected to USB power supply) when pressing the stop button in Pybricks Code to stop the program.

**Reply by dlech** [reply]
This crash is fixed by https://github.com/pybricks/pybricks-micropython/commit/1648303452cd4039f4f2384c7e89de3b8c44d9b5
